src/core/processing/qgsprocessingparametermeshdataset.cpp

```diff
@@ -121,11 +121,20 @@ QList<int> QgsProcessingParameterMeshDatasetGroups::valueAsDatasetGroup( const Q
     ret << 0;
   else
   {
-    QVariantList list = value.toList();
-    if ( list.isEmpty() )
-      ret << 0;
-    for ( const QVariant &v : list )
-      ret.append( v.toInt() );
+    if ( value.type() == QVariant::List )
+    {
+      //here we can't use  QgsProcessingParameters::parameterAsInts() because this method return empry list when first value is 0...
```

Comment:
hmm - that sounds like a bug to me!

Reply:
I strongly suspect it's a bug... likely it's been sitting there ever since the port from python or before! I'd rather see parameterAsInts fixed to be predictable and the individual raster algorithms handling "0" themselves if there's special handling required there.

Suffice to say, I'll accept the responsibility of regressions if you fix parameterAsInts and there is fallout from the fix! :+1:

Reply:
just one comment before fixing it, it really looks like intentional:
https://github.com/qgis/QGIS/blob/master/src/core/processing/qgsprocessingparameters.cpp#L283

Reply:
@nirvn looks like you wrote this code -- have you got any insights into why this logic is in place?

Reply:
@nyalldawson , I have unfortunately no recollection of what I did that. I vaguely remember an issue with string list -> ints list, but completely forgot the specifics. 

I think it's worth tweaking the parameterAsInts() here. The two scenarios I'd test with rearrange bands is checking that alg. properly handles empty band list and a band list with only the 1st band selected.

Reply:
ok, parameterAsInts fixed...
On my side, tests are ok.